RecoTracker/TkSeedGenerator/src/SeedingOTEDProducer.cc

```diff
@@ -0,0 +1,319 @@
+        for (auto vh : DSViter) {
+          VHsOnLayer.push_back(vh);
+        }
+      }
+    }
+  }
+
+  return VHsOnLayer;
+}
+
+void SeedingOTEDProducer::printVHsOnLayer(edm::Handle<VectorHitCollectionNew> VHs, unsigned int layerNumber) {
+  const VectorHitCollectionNew& input = *VHs;
+  if (!input.empty()) {
+    for (auto DSViter : input) {
+      for (auto vh : DSViter) {
+        if (checkLayer(DSViter.id()) == layerNumber)
+          std::cout << " VH in layer " << layerNumber << " >> " << vh << std::endl;
+      }
+    }
+  } else {
+    std::cout << " No VHs in layer " << layerNumber << "." << std::endl;
+  }
+}
+
+const TrajectoryStateOnSurface SeedingOTEDProducer::buildInitialTSOS(VectorHit& vHit) {
+  // having fun with theta
+  Global3DVector gv(vHit.globalPosition().x(), vHit.globalPosition().y(), vHit.globalPosition().z());
+  float theta = gv.theta();
+  // gv transform to local (lv)
+  const Local3DVector lv(vHit.det()->surface().toLocal(gv));
+
+  //FIXME::charge is fine 1 every two times!!
+  int charge = 1;
+  float p = vHit.momentum(magField);
+  float x = vHit.localPosition().x();
+  float y = vHit.localPosition().y();
+  float dx = vHit.localDirection().x();
+  // for dy use second component of the lv renormalized to the z component
+  float dy = lv.y() / lv.z();
+
+  // Pz and Dz should have the same sign
+  float signPz = copysign(1.0, vHit.globalPosition().z());
+
+  LocalTrajectoryParameters ltpar2(charge / p, dx, dy, x, y, signPz);
+  AlgebraicSymMatrix mat = assign44To55(vHit.parametersError());
+  // set the error on 1/p
+  mat[0][0] = pow(computeInverseMomentumError(vHit, theta, magField, beamSpot->sigmaZ()), 2);
+
+  //building tsos
+  LocalTrajectoryError lterr(asSMatrix<5>(mat));
+  const TrajectoryStateOnSurface tsos(ltpar2, lterr, vHit.det()->surface(), magField);
+
+  return tsos;
+}
+
+AlgebraicSymMatrix SeedingOTEDProducer::assign44To55(AlgebraicSymMatrix mat44) {
+  if (mat44.num_row() != 4 || mat44.num_col() != 4)
+    assert("Wrong dimension! This should be a 4x4 matrix!");
+
+  AlgebraicSymMatrix result(5, 0);
+  for (int i = 1; i < 5; i++) {
+    for (int j = 1; j < 5; j++) {
+      result[i][j] = mat44[i - 1][j - 1];
+    }
+  }
+  return result;
+}
+
+std::pair<bool, TrajectoryStateOnSurface> SeedingOTEDProducer::propagateAndUpdate(
+    const TrajectoryStateOnSurface initialTSOS, const Propagator& prop, const TrackingRecHit& hit) {
+  TrajectoryStateOnSurface propTSOS = prop.propagate(initialTSOS, hit.det()->surface());
+  TrajectoryStateOnSurface updatedTSOS = theUpdator->update(propTSOS, hit);
+  if
+    UNLIKELY(!updatedTSOS.isValid()) return std::make_pair(false, updatedTSOS);
+  return std::make_pair(true, updatedTSOS);
+}
+
+float SeedingOTEDProducer::computeGlobalThetaError(const VectorHit& vh, const double sigmaZ_beamSpot) {
+  double derivative =
+      vh.globalPosition().perp() / (pow(vh.globalPosition().z(), 2) + pow(vh.globalPosition().perp(), 2));
```

Comment:
check for divide by 0?

Reply:
I don't think both z() and perp() can ever be simultaneously 0 as no VH will ever be at (0,0,0).

Reply:
I agree